File: framework/test2/tree/mirror/mirror.go
package main

import (
    . "missingjs.com/soda/leetcode"
    "missingjs.com/soda/unittest"
    "missingjs.com/soda/util"
)

var logger = util.Logger
var _used ListNode

func mirror(root *TreeNode) *TreeNode {
    if root == nil {
        return nil
    }
    mirror(root.Left)
    mirror(root.Right)
    temp := root.Left
    root.Left = root.Right
    root.Right = temp
    return root
}

func main() {
    // create tester by function
    work := unittest.CreateWork(mirror)
    // OR create by struct factory
    // work := unittest.CreateWorkForStruct(Constructor)
    // work.SetValidator(func(e,r)bool)
    work.CompareSerial = true
    work.Run()
}

File: framework/test2/tree/mirror/mirror.ts
import {
    AvlTree,
    BinarySearchTree
} from '@datastructures-js/binary-search-tree';
import {
    PriorityQueue,
    MinPriorityQueue,
    MaxPriorityQueue
} from '@datastructures-js/priority-queue';
import { ListNode, NestedInteger, TreeNode } from 'soda/leetcode';
import { TestWork, Utils, Validators } from 'soda/unittest';

// step [1]: implement solution function
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
function mirror(root: TreeNode | null): TreeNode | null {
    if (!root) {
        return null;
    }
    mirror(root.left);
    mirror(root.right);
    [root.left, root.right] = [root.right, root.left];
    return root;
}

// step [2]: setup function/return/arguments
let taskFunc = mirror;
const work = TestWork.create<ReturnType<typeof taskFunc>>(taskFunc);
// let work = TestWork.forStruct(STRUCT);
// work.validator = (x, y) => { ... };
work.compareSerial = true;
console.log(work.run(Utils.fromStdin()));
